### DISCUSSION TITLE
Why spread-minimizing tokens could not work?

### DISCUSSION THREAD

**Discussion by deniszh** [grafana/mimir#9992] [General]
We're not using shuffle sharding yet, and I noticed quite severe memory disbalance in ingesters, so, we decided to migrate to spread-minimizing tokens. So, I did migration as described in https://grafana.com/docs/mimir/latest/configure/configure-spread-minimizing-tokens/ ... and nothing happens?
I still see same disbalance in cortex_ingester_memory_series (ignore pod name - only ingester deployed there)
<img width="1386" alt="Screenshot 2024-11-22 at 17 50 27" src="https://github.com/user-attachments/assets/e6f03ce5-239c-4df0-9c1c-2dacf8b404b9">

Writers and distributors have following flags enabled:
```
 -ingester.ring.spread-minimizing-join-ring-in-order=true
 -ingester.ring.spread-minimizing-zones=zone-a,zone-b,zone-c
 -ingester.ring.token-generation-strategy=spread-minimizing
 -ingester.ring.tokens-file-path=
```
Can it be because I was not able to remove `/data/tokens` at last step because distroless image now?
On the other hand disbalance didn't go after migrating first zone, so, don't think it's related.

**Comment by deniszh** [comment]
Still don't know what was the issue, but after rerolling cluster issue is gone 🤷 
![Screenshot 2024-11-28 at 15 28 26](https://github.com/user-attachments/assets/6670f08f-43ba-4240-bee5-d6cf0b4571b1)

**Reply by gespi1** [reply]
and yet the `/data/tokens` file was never removed?

**Reply by gespi1** [reply]
what do you mean specifically with `rerolling cluster` ?

**Reply by deniszh** [reply]
> and yet the /data/tokens file was never removed?

No, I didn't touch pvc


> what do you mean specifically with rerolling cluster ?

deleted all resources (through `kubectl delete -f <path to manifests generated from helm>`) and created them again.
please note that pvc were not affected.

**Reply by gespi1** [reply]
updated to spread-minimizing tokens on our dev environment and at first the number of series per ingester were not balanced but after a few hours they balanced themselves out.

**Reply by deniszh** [reply]
Yes, I also can't reproduce. But spreading-minimised tokens were definitely not working - now I see perfectly aligned percentages on ingester ring web ui, before it was not. So, let's assume it was some strange bug. 🤷

**Reply by pracucci** [reply]
> updated to spread-minimizing tokens on our dev environment and at first the number of series per ingester were not balanced but after a few hours they balanced themselves out.

As soon as you switch to spread-minimizing token (SMT), the series get reshuffled between ingesters. Ingesters will keep having the old (stale) series until they're compacted into a block, which may take up to 3h (TSDB head is compacted every 2h, but each compaction compacts the range [-3h ago, -1h ago]). So as soon as you change tokens you will see an increase in ingesters in-memory series due to the reshuffling and you will still see unbalanced series per ingesters, then after up to 3h the series should stabilize and they should be better balanced between ingesters.

**Reply by deniszh** [reply]
Hi Marco,
You're totally right - that how spread-minimizing token  should work. And for some reason it didn't - but I can't reproduce that bug anymore, and it's not the case for us - so, I think we can close that discussion.
Thanks for reply, if I find more info - then I reopen this.
